File: GalaxyMonkey/AnimationCatalog.swift
//
//  AnimationCatalog.swift
//  GalaxyMonkey
//
//  Loads SKTextureAtlas-backed frame sequences from `.spriteatlas` folders
//  in Assets.xcassets. Mirrors SpriteCatalog's caching + nil-on-missing
//  pattern so callers can branch gracefully when the art isn't present.
//
//  Frames are alpha-sorted so they play in numeric file order
//  (e.g. gorilla_windup_01..08).
//

import SpriteKit

enum AnimationSet: String, CaseIterable {
    case gorillaIdle   = "GorillaIdleAnim"
    case gorillaWindup = "GorillaWindupAnim"
    case bombExplosion = "BombExplosionAnim"
    case droneSwarmIdle           = "DroneSwarmIdleAnim"
    case plasmaJellyIdle          = "PlasmaJellyIdleAnim"
    case preppyLeftIdle           = "PreppyLeftIdleAnim"
    case preppyRightIdle          = "PreppyRightIdleAnim"
    case whiteLeftIdle            = "WhiteLeftIdleAnim"
    case whiteRightIdle           = "WhiteRightIdleAnim"
    case shadyLeftIdle            = "ShadyLeftIdleAnim"
    case shadyRightIdle           = "ShadyRightIdleAnim"
    case heavyCosmonautLeftIdle   = "HeavyCosmonautLeftIdleAnim"
    case heavyCosmonautRightIdle  = "HeavyCosmonautRightIdleAnim"
    case astroSniperLeftIdle      = "AstroSniperLeftIdleAnim"
    case astroSniperRightIdle     = "AstroSniperRightIdleAnim"
    case miniBossLeftIdle         = "MiniBossLeftIdleAnim"
    case miniBossRightIdle        = "MiniBossRightIdleAnim"
    // Walk cycles. Used while the enemy is in motion; idle plays only when
    // the enemy is effectively stationary (e.g. attacking from range).
    case preppyLeftWalk           = "PreppyLeftWalkAnim"
    case preppyRightWalk          = "PreppyRightWalkAnim"
    case whiteLeftWalk            = "WhiteLeftWalkAnim"
    case whiteRightWalk           = "WhiteRightWalkAnim"
    case shadyLeftWalk            = "ShadyLeftWalkAnim"
    case shadyRightWalk           = "ShadyRightWalkAnim"
    case heavyCosmonautLeftWalk   = "HeavyCosmonautLeftWalkAnim"
    case heavyCosmonautRightWalk  = "HeavyCosmonautRightWalkAnim"
    case astroSniperLeftWalk      = "AstroSniperLeftWalkAnim"
    case astroSniperRightWalk     = "AstroSniperRightWalkAnim"
    // Mini-boss walk is a front-3/4 stomp with no lateral motion — one
    // atlas serves both facings.
    case miniBossWalk             = "MiniBossWalkAnim"
}

enum AnimationCatalog {

    private static var cache: [AnimationSet: [SKTexture]] = [:]
    private static var missing: Set<AnimationSet> = []

    /// Returns the ordered frame list. Empty array if the atlas is missing
    /// or contains no frames — call sites should check `isEmpty`.
    static func textures(for set: AnimationSet) -> [SKTexture] {
        if let cached = cache[set] { return cached }
        if missing.contains(set) { return [] }

        let atlas = SKTextureAtlas(named: set.rawValue)
        let names = atlas.textureNames.sorted()
        guard !names.isEmpty else {
            missing.insert(set)
            return []
        }
        let textures = names.map { atlas.textureNamed($0) }
        for t in textures { t.filteringMode = .linear }
        cache[set] = textures
        return textures
    }

    static func hasTextures(_ set: AnimationSet) -> Bool {
        !textures(for: set).isEmpty
    }

    /// Convenience: build an animate-then-remove action for a one-shot fx.
    static func oneShot(_ set: AnimationSet, frameDuration: TimeInterval) -> SKAction? {
        let frames = textures(for: set)
        guard !frames.isEmpty else { return nil }
        return SKAction.sequence([
            SKAction.animate(with: frames, timePerFrame: frameDuration),
            SKAction.removeFromParent(),
        ])
    }

    /// Convenience: build a forever-repeating action (for idle loops).
    static func loop(_ set: AnimationSet, frameDuration: TimeInterval) -> SKAction? {
        let frames = textures(for: set)
        guard !frames.isEmpty else { return nil }
        return SKAction.repeatForever(SKAction.animate(with: frames, timePerFrame: frameDuration))
    }
}

File: android/core/src/main/kotlin/dev/copt/galaxymonkey/AnimationCatalog.kt
package dev.copt.galaxymonkey

import com.badlogic.gdx.graphics.g2d.Animation
import com.badlogic.gdx.graphics.g2d.Animation.PlayMode
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.utils.Array as GdxArray

// Mirrors SpriteCatalog's caching + null-on-missing pattern so callers
// branch gracefully when art isn't present yet. Frames are loaded via
// atlas.findRegions which returns them in index order (matching iOS
// textureNames.sorted() ordering).

enum class AnimationSet(val atlasBase: String) {
    GORILLA_LEFT_IDLE("gorilla_left_idle"),
    GORILLA_RIGHT_IDLE("gorilla_right_idle"),
    GORILLA_LEFT_WINDUP("gorilla_left_windup"),
    GORILLA_RIGHT_WINDUP("gorilla_right_windup"),
    BOMB_EXPLOSION("bomb_explosion"),
    DRONE_SWARM_LEFT_IDLE("drone_swarm_left_idle"),
    DRONE_SWARM_RIGHT_IDLE("drone_swarm_right_idle"),
    PLASMA_JELLY_IDLE("plasma_jelly_idle"),
    PREPPY_LEFT_IDLE("preppy_left_idle"),
    PREPPY_RIGHT_IDLE("preppy_right_idle"),
    WHITE_LEFT_IDLE("white_left_idle"),
    WHITE_RIGHT_IDLE("white_right_idle"),
    SHADY_LEFT_IDLE("shady_left_idle"),
    SHADY_RIGHT_IDLE("shady_right_idle"),
    HEAVY_COSMONAUT_LEFT_IDLE("heavy_cosmonaut_left_idle"),
    HEAVY_COSMONAUT_RIGHT_IDLE("heavy_cosmonaut_right_idle"),
    ASTRO_SNIPER_LEFT_IDLE("astro_sniper_left_idle"),
    ASTRO_SNIPER_RIGHT_IDLE("astro_sniper_right_idle"),
    MINI_BOSS_LEFT_IDLE("mini_boss_left_idle"),
    MINI_BOSS_RIGHT_IDLE("mini_boss_right_idle"),
    PREPPY_LEFT_WALK("preppy_left_walk"),
    PREPPY_RIGHT_WALK("preppy_right_walk"),
    WHITE_LEFT_WALK("white_left_walk"),
    WHITE_RIGHT_WALK("white_right_walk"),
    SHADY_LEFT_WALK("shady_left_walk"),
    SHADY_RIGHT_WALK("shady_right_walk"),
    HEAVY_COSMONAUT_LEFT_WALK("heavy_cosmonaut_left_walk"),
    HEAVY_COSMONAUT_RIGHT_WALK("heavy_cosmonaut_right_walk"),
    ASTRO_SNIPER_LEFT_WALK("astro_sniper_left_walk"),
    ASTRO_SNIPER_RIGHT_WALK("astro_sniper_right_walk"),
    MINI_BOSS_WALK("mini_boss_walk"),
}

object AnimationCatalog {

    private lateinit var atlas: TextureAtlas
    private val cache = mutableMapOf<AnimationSet, GdxArray<TextureAtlas.AtlasRegion>>()
    private val missing = mutableSetOf<AnimationSet>()

    fun init(atlas: TextureAtlas) {
        this.atlas = atlas
        cache.clear()
        missing.clear()
    }

    fun frames(set: AnimationSet): GdxArray<TextureAtlas.AtlasRegion> {
        cache[set]?.let { return it }
        if (set in missing) return GdxArray(0)
        val regions = atlas.findRegions(set.atlasBase)
        if (regions.size == 0) {
            missing.add(set)
            return regions
        }
        cache[set] = regions
        return regions
    }

    fun has(set: AnimationSet): Boolean = frames(set).size > 0

    fun oneShot(set: AnimationSet, frameDuration: Float): Animation<TextureRegion>? {
        val f = frames(set)
        if (f.size == 0) return null
        return Animation(frameDuration, f, PlayMode.NORMAL)
    }

    fun loop(set: AnimationSet, frameDuration: Float): Animation<TextureRegion>? {
        val f = frames(set)
        if (f.size == 0) return null
        return Animation(frameDuration, f, PlayMode.LOOP)
    }

    fun totalDuration(animation: Animation<TextureRegion>): Float =
        animation.frameDuration * animation.keyFrames.size
}
